### PR TITLE
Fixed issue #18562: Upload question type partially broken

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Review by Shnoulle — CHANGES_REQUESTED** [review]
This don't fix the major issue.

When move previous : uploaded files disappear

And about `fieldname` test : it was here for security : disable robot load (if htaccess is used). then we must send it in URL.

**Comment by Shnoulle** [comment]
The biggest issue are not fixed 
Peek 22-12-2022 17-57
Peek 22-12-2022 17-59


This is this issue that was related to this commit : https://github.com/LimeSurvey/LimeSurvey/commit/cc70269d03206cba6683367fa8e9a6e56c72a71a

File: application/controllers/UploaderController.php [hunk 1]
```diff
@@ -38,10 +38,13 @@ public function run($actionID)
         $sFileGetContent = Yii::app()->request->getParam('filegetcontents', ''); // The file to view fu_ or fu_tmp
         $bDelete = Yii::app()->request->getParam('delete');
         $sFieldName = Yii::app()->request->getParam('fieldname');
-        $aFieldMap = createFieldMap($oSurvey, 'short', false, false, $sLanguage);
-        if (!isset($aFieldMap[$sFieldName])) {
-            throw new CHttpException(400); // See for debug > 1
-        }        
+        // Make sure the fieldname is valid, but only when it's needed (it's not needed on 'filegetcontents' requests)
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/controllers/UploaderController.php:41]: > This was added for security issue 

I dropped some comments on mantis about it.
There are 4 actions on same endpoint.
I guess the security check wasn't for getting a temp file with a random file name, but for the others actions

**Comment by gabrieljenik** [comment]
> When move previous : uploaded files disappear

OK, this is new info

**Comment by Shnoulle** [comment]
> > When move previous : uploaded files disappear
> 
> OK, this is new info

It was in the original issue … 
Capture d’écran du 2022-12-22 18-06-43

**Comment by gabrieljenik** [comment]
> > > When move previous : uploaded files disappear
> > 
> > 
> > OK, this is new info
> 
> It was in the original issue … Capture d’écran du 2022-12-22 18-06-43

image

File: application/controllers/UploaderController.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/controllers/UploaderController.php:41]: > I guess the security check wasn't for getting a temp file with a random file name, but for the others actions

Files can contain confidential content … have a session check is really a great solution. Robot on multiple serve r: you can get a lot of confidential files. We must do **all** we can to disallow such attack.

If you want : i can take this part if you can fix the pint 2.

**Comment by Shnoulle** [comment]
Bug #2 #2291 
But right : next time i create 2 issues …

File: application/controllers/UploaderController.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/controllers/UploaderController.php:41]: > Files can contain confidential content … have a session check is really a great solution. Robot on multiple serve r: you can get a lot of confidential files. We must do all we can to disallow such attack.

This seems an enhancement, not a bug fix.

The error here was that the fieldname was not sent.
Hence validation failed.
We can:
a- Send the fieldname
b- Disable validation for this action

A is not more secure than B, as getting the fieldname is probably the easiest part.
Still, we can add it if it makes things more secure

**Comment by gabrieljenik** [comment]
Check now @Shnoulle please.
- Restored the validation
- Added the fieldname to the urls
- Applied some fixes on the EM part

File: application/controllers/UploaderController.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/controllers/UploaderController.php]: > Create session need a $_POST action, here it's a $_GET

Session should be already created. If not upload controller will not process.
https://github.com/LimeSurvey/LimeSurvey/blob/6af1ced564ef49165726a45a971d96b0ec4ffe5c/application/controllers/UploaderController.php#L23

Anyway, we added the fieldname to the url

File: assets/scripts/modaldialog.js [hunk 2]
```diff
@@ -118,7 +118,7 @@ function displayUploadedFiles(jsonstring, filecount, fieldname, show_title, show
         for (i = 0; i < filecount; i++)
         {
             if (isValueInArray(image_extensions, jsonobj[i].ext))
-                display += '<tr><td class="upload image"><img src="' + uploadurl + '/filegetcontents/'+decodeURIComponent(jsonobj[i].filename)+'" class="uploaded" /></td>';
+                display += '<tr><td class="upload image"><img src="' + uploadurl + '/filegetcontents/'+decodeURIComponent(jsonobj[i].filename+'/fieldname/'+fieldname)+'" class="uploaded" /></td>';
```
Comment (Shnoulle) [inline assets/scripts/modaldialog.js:121]: This url need to be tested with GET and PATH url

**Review by Shnoulle — APPROVED** [review]
Thank you
Not tested

File: assets/scripts/modaldialog.js (hunk 2 above)
Comment (gabrieljenik) [reply assets/scripts/modaldialog.js:121]: Doesn't it work right now on both? what would be different now?
Still, yes, we could test on both

Comment (Shnoulle) [reply assets/scripts/modaldialog.js:121]: GET URL index.php?r=uploader/index/mode/upload/filegetcontents/fu_ickrap62w67ze4f/fieldname/136931X1X2

We need to review when we rewrite this controller :)

**Comment by olleharstedt** [comment]
Both tested and needs testing? Tag confusion.

**Comment by gabrieljenik** [comment]
Oops, I went thorugh this yesteday and thought the tested tag was there. :)

**Comment by Shnoulle** [comment]
I test it …
It's really a broken issue here.

**Comment by gabrieljenik** [comment]
> I test it … It's really a broken issue here.

image

**Comment by gabrieljenik** [comment]
Is this tested by you Denis? 
We have tested it as well, but as this is EM related and seems you have tested it, I reask before setting it to tested ok.

**Comment by Shnoulle** [comment]
> Is this tested by you Denis? We have tested it as well, but as this is EM related and seems you have tested it, I reask before setting it to tested ok.

I test it too, on another instance than mine (apply fix on a release) + mine test instance

**Comment by olleharstedt** [comment]
Good work here!

**Pull request merged**